### DISCUSSION TITLE
Different codes "captured" by IrScrutinizer and HomeAssistant Broadlink 'learn' - I'm confused.

### DISCUSSION THREAD

**Discussion by KDYR** [bengtmartensson/IrScrutinizer#479] [Q&A]
Thanks a lot to @bengtmartensson for making such sophisticated and feature-rich tool available! That's awesome!

Ultimately I would like to generate and send IR codes in the Denon-K format based on info I have on device, sub-device and function numbers; I'd like to generate and send codes for functions that are not available on the std remote, so I cannot just grab the codes from there.
Now, trying to learn I have built a Girs device and installed IrScrutinizer; now I am trying things out: My remote has a button "Tuner", and using IrScrutinizer to capture and analyze, I get:

`Denon-K: {D=4,S=1,F=719}, beg=0, end=99, reps=1`

...where "D=4,S=1,F=719" corresponds to info in my pdf-file with functions. So far, so good.

I can then Export the code to a yaml file using the "Homeass-Broadlink IR" -format, and I get this code:

`JgBkAGwqDQ0NDQ0qDQ0NKg0NDSoNDQ0NDSoNDQ0NDSoNKg0NDQ0NDQ0NDQ0NDQ0NDQ0NKg0NDSoNDQ0NDQ0NKg0qDSoNKg0NDQ0NKg0qDQ0NKg0NDQ0NKg0NDSoNKg0qDQ0NDQ0qDQANBQ==`

Now, I can also get the code using my Broadlink RM4pro, HomeAssistant developer tools, and 'learn', but then I get something that is different, and longer:

`JgDMAGo4Dg4NDg0rDA8OKQ4ODSoNDwwPDSoODg0PDCsNKg0PDQ4ODgwQDQ4ODg0ODg4OKQ4ODCsNDw0ODg4NKg0qDikOKQ4ODQ8MKw0qDQ8NKg0ODg4MKw0PDSoNKg0qDQ8MDw4pDQAJbWs5DQ4ODgwrDQ8MKw0ODSoODg0PDSoNDwwPDikOKQ4ODg4NDg4ODQ8NDg0PDQ8NKgwPDSoNDw4ODA8NKw0qDSoNKg0PDA8OKQ4pDg4NKg4ODBAMKw0ODSoNKg4qDQ4NDw0qDQANBQ==`

Same remote, same button pushed, but different results. What is it that I failed to understand? I am concerned that IrScrutinizer generated codes are not (fully) usable in HA, when HA sees a button differently from IrScrutinizer.

Anyone who can see through this and point me in the right direction?

**Comment by bengtmartensson** [comment]
Sorry, I am not able to help right how. Will get back to you towards the end of the week.

In the meantime, please check my pending PR for the python-broadlink project.

**Reply by bengtmartensson** [reply]
See if you can checkout P.R741 in pytbon-broadlink and use its - - convert. I am I  hospital right now.

**Reply by KDYR** [reply]
THANKS for you message! You should probably focus on getting well.
Later we may pick up on this - I do not have the knowledge required to understand your suggestion.
BR

**Comment by KDYR** [comment]
Tried to follow the instruction for the python-broadlink and got a code in a different format:

![image](https://user-images.githubusercontent.com/16138808/217100493-6e3e5602-0b98-4dad-8650-a95bb45ee964.png)

Code is:

`b'&\x00\xcc\x00k7\x0e\x0e\r\x0f\r*\r\x0e\x0e)\x0e\x0e\x0e)\x0e\x0e\r\x0f\x0c+\r\x0e\r\x0f\x0c+\r*\r\x0f\x0c\x0f\x0e\x0e\r\x0f\x0c\x0f\r\x0f\r\x0f\r\x0e\x0e)\r\x0f\r*\x0e\x0e\x0c\x0f\x0e\x0e\r*\r*\x0e)\x0e)\x0e\x0e\r\x0f\x0c+\x0c+\r\x0f\r*\x0c\x0f\r\x0f\r*\r\x0f\r*\r*\r*\r\x0f\r\x0e\r*\x0e\x00\tkm7\r\x0f\r\x0e\x0e)\x0e\x0e\r*\x0e\x0e\r*\r\x0e\x0e\x0e\x0e)\x0e\x0e\x0c\x0f\x0e*\r*\r\x0e\x0e\x0e\r\x0f\x0c\x0f\x0e\x0e\x0c\x10\x0c\x0f\r\x0f\r*\r\x0f\x0c+\r\x0e\x0e\x0e\x0c\x0f\x0e*\r*\r*\r*\r\x0f\r\x0e\r*\x0e)\x0e\x0e\r*\x0e\x0e\r\x0e\x0e*\r\x0e\r*\x0e)\x0e)\r\x0f\r\x0f\r*\r\x00\r\x05'`

Not really sure if this fits with the IrScrutinizer captured code or the HomeAssistant captured code.

**Comment by KDYR** [comment]
Using instead [Tomer Klein's](https://github.com/t0mer) Windows Server [Broadlink Manager](https://bit.ly/BM_Win) (which is implementing [Matthew Garrett's Python-Broadlink](https://github.com/mjg59/python-broadlink)) I get this:

`JgDMAGo5DQ4NDw0qDQ8NKg0ODikODgwQDCsNDg4ODSoNKg0PDQ4ODg0PDA8ODg0ODQ8NKg4ODCsNDw0ODQ8NKg0qDSoNKg4ODQ4NKw0qDQ4OKQ4ODQ8MKw0ODSoOKQ0qDQ8ODgwrDQAJam43DQ4ODg0qDQ8MKw0ODikODg4ODCsNDg0PDSoOKQ4ODBAMDw4ODQ4NDw0PDQ4OKQ4ODikNDw0ODg4NKg4pDikOKQ4ODQ8MKw0qDQ8NKg0ODg4NKg0PDSoNKg0qDQ8NDg0qDQANBQAAAAAAAAAAAAA=`

...which is different from codes grabbed with IrScrutinizer and HomeAssistant .

**Comment by bengtmartensson** [comment]
So, now I am back again!!

An IR-signal is essentially a number of durations, in micro seconds. The "Broadlink format" stores them slightly coded. The said export format converts the durations in IrScrutinizer to the Broadlink format. Also, I wrote a function to do this conversion (both directions) for python-broadlink, triggered by the `--convert` option to the broadlink_cli program, about two years ago. Unfortunately (developments of the Python language?), it broke, and I also discovered some problems. So I fixed it, and submitted [a new PR]( https://github.com/mjg59/python-broadlink/pull/741) (a month ago or so). Try to check out this version. GitHub has many good guides on how to do such things.

Also, the current [snapshot](https://github.com/bengtmartensson/IrScrutinizer/releases/tag/ci-build) can understand Broadlink format, pasted into the Scrutinize signal window. It also has an extended Text export, that can generate Broadlink.

> I can then Export the code to a yaml file using the "Homeass-Broadlink IR" -format

Unfortunately, this is slight broken. I plan to fix it coming week. Keep a look at https://github.com/bengtmartensson/IrScrutinizer/issues/477 . 

> Same remote, same button pushed, but different results. What is it that I failed to understand?

First, the durations are empirically measured times, so a certain fluctuation is normal. Secondly, when you hold down a key, (some parts of) the signal is repeated. IrScrutinizer has a very advanced function ("repeat finder") that finds an eliminates the repeats. This explains that the signals you have learned with the "dumb" devices are much longer, since they do not contain repeat elimination. Just paste you signals into the (snapshot version) of IrScrutinizer, and it will reduce it.

Hope this helps!
